### PR TITLE
Fixed #24509 -- Added Expression support to SQLInsertCompiler

### PR DESCRIPTION
**Pull request by jarshwah**
This is an update/rebase/docs to: https://github.com/django/django/pull/5099

### CODE REVIEW

**Comment by jarshwah** [comment]
I'm not too sure if this is ready to be merged mainly due to the comment on the original patch: https://github.com/django/django/pull/5099#discussion-diff-36384527

If @akaariai or @AlexHill are able to say that we can merge without addressing this comment, and docs are cleared as being OK, then we should be able to get this in.

**Comment by akaariai** [comment]
Field.get_placeholder() is IIRC a hack to make Oracle GIS work. I wouldn't worry too much about it. We want to get rid of it, but we don't have to do that in this patch.

File: docs/ref/models/expressions.txt [hunk 1]
```diff
@@ -5,10 +5,14 @@ Query Expressions
...
-Expressions can be combined, or in some cases nested, to form more complex
-computations.
+an update, create, filter, order by, annotation, or aggregate. There are a
+number of built-in expressions (documented below) that can be used to help you
+write queries. Expressions can be combined, or in some cases nested, to form
+more complex computations.
+
+.. versionchanged:: 1.9
+
+    Support for using expressions with new model instances.
```
Comment (jarshwah) [inline docs/ref/models/expressions.txt]: Support for using expressions when creating new model instances.

**Comment by timgraham** [comment]
buildbot, test on oracle.

File: django/db/models/sql/compiler.py [hunk 2]
```diff
@@ -909,17 +909,107 @@ def __init__(self, *args, **kwargs):
...
+            value = value.resolve_expression(self.query, allow_joins=False, for_save=True)
+            # Don't allow values containing Col expressions. They refer to existing
+            # columns on a row, but in the case of insert the row doesn't exist yet.
+            if value.contains_column_references:
+                raise ValueError(
+                    'Failed to insert expression "%s" on %s. F() expressions '
+                    'can only be used to update, not to insert.' % (value, field)
+                )
+            if value.contains_aggregate:
+                raise FieldError("Aggregate functions are not allowed in this query")
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: missing test for this error?

**Comment by timgraham** [comment]
The docs seem a bit sparse. A few examples might help? Is it common knowledge, for example, that you must use `Value()` to wrap literals inside expressions? The docs say, "You will rarely need to use ``Value()`` directly." Should that be revised in light of this? Having to refresh the model to get the values that the database could also be a point worth mentioning. Is a test for the `.objects.create()` version worthwhile (not sure how similar the code path is to `.save()`? 

Cosmetic edits: http://dpaste.com/1TA06CQ

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by jarshwah** [comment]
I agree the docs for this specific support is lacking. I've added a few more examples and tidied up some of the concerns you had. I think the documentation here could benefit from a restructuring, so I've opened another ticket to track that: https://code.djangoproject.com/ticket/25444

File: django/db/models/sql/compiler.py [hunk 3]
```diff
@@ -933,35 +1018,27 @@ def as_sql(self):
...
-                    ) for f in fields
-                ]
+            value_rows = [
+                [self.prepare_value(field, self.pre_save_val(field, obj)) for field in fields]
                 for obj in self.query.objs
             ]
         else:
-            values = [[self.connection.ops.pk_default_value()] for obj in self.query.objs]
-            params = [[]]
+            # I guess this case is just for an entirely empty object.
```
Comment (timgraham) [inline django/db/models/sql/compiler.py]: This comment could be improved or removed.

**Comment by timgraham** [comment]
edits: http://dpaste.com/1B0CSF2
LGTM otherwise.

**Pull request merged**